tests/xdc/stream_based_replication_test.go

```diff
@@ -0,0 +1,379 @@
+			}
+			historyBatch = append(historyBatch, historyEvents)
+		}
+
+		versionHistory := s.eventBatchesToVersionHistory(nil, historyBatch)
+		s.importEvents(
+			workflowID,
+			runID,
+			versionHistory,
+			historyBatch,
+			historyClient,
+			true,
+		)
+
+		executions = append(executions, &commonpb.WorkflowExecution{WorkflowId: workflowID, RunId: runID})
+	}
+	return executions
+}
+
+func (s *streamBasedReplicationTestSuite) waitUntilWorkflowReplicated(
+	ctx context.Context,
+	namespace string,
+	workflowId string,
+	runId string,
+) error {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			res, err := s.cluster2.GetAdminClient().DescribeMutableState(ctx, &adminservice.DescribeMutableStateRequest{
+				Namespace: namespace,
+				Execution: &commonpb.WorkflowExecution{
+					WorkflowId: workflowId,
+					RunId:      runId,
+				},
+			})
+			if err != nil {
+				return err
+			}
+			if res.GetDatabaseMutableState().GetExecutionState().GetStatus() != enums.WORKFLOW_EXECUTION_STATUS_RUNNING {
+				return nil
+			}
+		}
+	}
+}
+
+func (s *streamBasedReplicationTestSuite) waitUntilNamespaceReplicated(
+	ctx context.Context,
+	namespace string,
+) error {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			_, err := s.cluster2.GetFrontendClient().DescribeNamespace(ctx, &workflowservice.DescribeNamespaceRequest{
+				Namespace: namespace,
+			})
+			if err != nil {
+				continue
+			}
+			return nil
+		}
+	}
+}
+
+func (s *streamBasedReplicationTestSuite) assertHistoryEvents(
+	ctx context.Context,
+	namespaceId string,
+	workflowId string,
+	runId string,
+) error {
+	mockClientBean := client.NewMockBean(s.controller)
+	mockClientBean.
+		EXPECT().
+		GetRemoteAdminClient("cluster1").
+		Return(s.cluster1.GetAdminClient(), nil).
+		AnyTimes()
+	mockClientBean.EXPECT().GetRemoteAdminClient("cluster2").Return(s.cluster2.GetAdminClient(), nil).AnyTimes()
+
+	serializer := serialization.NewSerializer()
+	cluster1Fetcher := replication.NewHistoryPaginatedFetcher(
+		nil,
+		mockClientBean,
+		serializer,
+		nil,
+		s.logger,
+	)
+	cluster2Fetcher := replication.NewHistoryPaginatedFetcher(
+		nil,
+		mockClientBean,
+		serializer,
+		nil,
+		s.logger,
+	)
+	iterator1 := cluster1Fetcher.GetSingleWorkflowHistoryPaginatedIterator(
+		ctx, "cluster1", namespace.ID(namespaceId), workflowId, runId, 0, 1, 0, 0)
+	iterator2 := cluster2Fetcher.GetSingleWorkflowHistoryPaginatedIterator(
+		ctx, "cluster2", namespace.ID(namespaceId), workflowId, runId, 0, 1, 0, 0)
+	for iterator1.HasNext() {
+		s.True(iterator2.HasNext())
+		batch1, err := iterator1.Next()
+		s.NoError(err)
+		batch2, err := iterator2.Next()
+		s.NoError(err)
+		s.Equal(batch1.VersionHistory.Items, batch2.VersionHistory.Items)
+		s.Equal(batch1.RawEventBatch, batch2.RawEventBatch)
+	}
+	return nil
```

Comment:
maybe add check iterator2.HasNext is false.

Reply:
Sure. Good idea.